src/quota/quota_parser.rs

```diff
@@ -0,0 +1,214 @@
+// license that can be found in the LICENSE file.
+
+use crate::{
+    base_parsers::{param, period, target},
+    quota::QMT,
+    types::{Param, Record, Target, TargetStats},
+    QuotaKind, QuotaStat, QuotaStatOsd, QuotaStats, TargetQuotaStat,
+};
+use combine::{
+    attempt, choice, eof,
+    error::{ParseError, StreamError},
+    many, one_of, optional,
+    parser::{
+        char::{alpha_num, newline, string},
+        repeat::take_until,
+    },
+    stream::{Stream, StreamErrorFor},
+    token, Parser,
+};
+
+pub(crate) const USR_QUOTAS: &str = "usr";
+pub(crate) const PRJ_QUOTAS: &str = "prj";
+pub(crate) const GRP_QUOTAS: &str = "grp";
+pub(crate) const QMT_STATS: [&str; 3] = [USR_QUOTAS, PRJ_QUOTAS, GRP_QUOTAS];
+
+/// Takes QMT_STATS and produces a list of params for
+/// consumption in proper ltcl get_param format.
+pub(crate) fn params() -> Vec<String> {
+    QMT_STATS
+        .iter()
+        .map(|x| format!("{QMT}.*.*.glb-{x}"))
+        .collect()
+}
+
+/// Parses a target name
+pub(crate) fn qmt_pool<I>() -> impl Parser<I, Output = (Target, Target)>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (
+        choice((
+            string("md").skip(token('-')).map(|x| Target(x.to_string())),
+            string("dt").skip(token('-')).map(|x| Target(x.to_string())),
+        )),
+        target(),
+    )
+}
+
+/// Parses the name of a target
+fn qmt_target<I>() -> impl Parser<I, Output = (Target, Target, Target)>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (target().skip(period()), qmt_pool().skip(period()))
+        .map(|(target, (manager, pool))| (target, manager, pool))
+        .message("while parsing target_name")
+}
+
+pub(crate) fn quota_stats<I>() -> impl Parser<I, Output = Vec<QuotaStat>>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (
+        optional(newline()), // If quota stats are present, the whole yaml blob will start on a newline
+        many::<Vec<_>, _, _>(alpha_num().or(one_of("_-:".chars()))), // But yaml header might not be indented, ignore it
+        newline(),
+        take_until(attempt((newline(), alpha_num()).map(drop).or(eof()))),
+    )
+        .skip(optional(newline()))
+        .skip(optional(eof()))
+        .and_then(|(_, _, _, x): (_, _, _, String)| {
+            serde_yaml::from_str::<Vec<QuotaStat>>(&x).map_err(StreamErrorFor::<I>::other)
+        })
+}
+
+pub(crate) fn quota_stats_osd<I>() -> impl Parser<I, Output = Vec<QuotaStatOsd>>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (
+        optional(newline()), // If quota stats are present, the whole yaml blob will start on a newline
+        many::<Vec<_>, _, _>(alpha_num().or(one_of("_-:".chars()))), // But yaml header might not be indented, ignore it
```

Comment:
We take a line of YAML here, but accept a subset of valid YAML...

This looks ok for now, but is a potential source of breakage going forwards.